lib/url.js

```diff
@@ -75,355 +88,192 @@ const slashedProtocol = {
-    if (parseQueryString) {
-      this.query = querystring.parse(this.query);
-    }
-    rest = rest.slice(0, qm);
-  } else if (parseQueryString) {
-    // no query string, but parseQueryString still requested
-    this.search = '';
-    this.query = {};
-  }
-  if (rest) this.pathname = rest;
-  if (slashedProtocol[lowerProto] &&
-      this.hostname && !this.pathname) {
+  if (!this.pathname && this.hostname && _slashProtocols[this._protocol])
     this.pathname = '/';
-  }
 
-  //to support http.request
-  if (this.pathname || this.search) {
-    var p = this.pathname || '';
-    var s = this.search || '';
-    this.path = p + s;
+  if (parseQueryString) {
+    var search = this.search;
+    if (search == null)
+      search = this.search = '';
+    if (search.charCodeAt(0) === 0x3F /*'?'*/)
+      search = search.slice(1);
+    // This calls a setter function, there is no .query data property.
+    this.query = querystring.parse(search);
   }
+};
+
+function urlResolve(source, relative) {
+  return urlParse(source, false, true).resolve(relative);
+}
 
-  // finally, reconstruct the href based on what has been validated.
-  this.href = this.format();
-  return this;
+Url.prototype.resolve = function(relative) {
+  return this.resolveObject(urlParse(relative, false, true)).format();
 };
 
-// format a parsed object into a url string
+// Format a parsed object into a url string.
 function urlFormat(obj) {
-  // ensure it's an object, and not a string url.
+  // Ensure it's an object, and not a string url.
   // If it's an obj, this is a no-op.
   // this way, you can call url_format() on strings
   // to clean up potentially wonky urls.
-  if (typeof obj === 'string') obj = urlParse(obj);
-
-  else if (typeof obj !== 'object' || obj === null)
+  if (typeof obj === 'string') {
+    obj = urlParse(obj);
+  } else if (typeof obj !== 'object' || obj === null) {
     throw new TypeError("Parameter 'urlObj' must be an object, not " +
                         obj === null ? 'null' : typeof obj);
-
-  else if (!(obj instanceof Url)) return Url.prototype.format.call(obj);
+  } else if (!(obj instanceof Url)) {
+    return Url.prototype.format.call(obj);
+  }
 
   return obj.format();
 }
 
 Url.prototype.format = function() {
   var auth = this.auth || '';
+
   if (auth) {
     auth = encodeURIComponent(auth);
     auth = auth.replace(/%3A/i, ':');
     auth += '@';
   }
 
-  var protocol = this.protocol || '',
-      pathname = this.pathname || '',
-      hash = this.hash || '',
-      host = false,
-      query = '';
-
-  if (this.host) {
-    host = auth + this.host;
-  } else if (this.hostname) {
-    host = auth + (this.hostname.indexOf(':') === -1 ?
-        this.hostname :
-        '[' + this.hostname + ']');
-    if (this.port) {
-      host += ':' + this.port;
-    }
+  var protocol = this.protocol || '';
+  var pathname = this.pathname || '';
+  var hash = this.hash || '';
+  var search = this.search || '';
```

Comment:
Isn't the `||` the same condition as on line 244. In my understanding 244 makes the `|| ''` part useless?!